Domains: allow different variations of root domain for supported records

klimeryk opened the pull request:
Before, we only supported an empty name (subdomain) as means of inputting the root domain. Since some users might by mistake or out of habit input other variations, this change will treat the following the same as empty subdomain and add a root domain (for records that it's allowed):
* `@`
* `@.example.com`
* `@.example.com.`
* `example.com`
* `example.com.`

This fixes #3318

/cc
@umurkontaci @aidvu for code review
@BandonRandon for additional testing

stephanethomas commented on client/lib/domains/dns/index.js:
```diff
@@ -5,22 +5,23 @@ const includes = require( 'lodash/collection/includes' ),
 			name: fieldName,
 			value: value,
 			type: fieldValues.type,
+			domainName: domainName
```
Note you could use a [shorthand](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Property_definitions) for this property name.

klimeryk replied:
I thought about it at first too, but coming from languages like Java, it feels weird to me to rearrange the parameters to a function like that (because our guidelines require shorthands to be at the top of the list). But I guess not using a shorthand could also be confusing, because someone could think that I'm mapping `domainName` to something else than `domainName`.
So shorthand it is! :laughing:

stephanethomas replied:
I must admit I'm also not particularly fond of enforcing shorthands to be listed first.

BandonRandon commented:
Patch works great!

Just a note that we may not need to verify for:
`@.example.com
@.example.com.`
However, I think it's fine if we under validate in this case :)

klimeryk commented:
Rebased to pull in changes related to lodash (I wanted to test if everything is still working correctly). I've fixed the issue @stephanethomas pointed while I was at it.

@BandonRandon: yeah, I know it's a bit of an overkill, but it doesn't costs us close to nothing in terms of code or performance. And given the volume of our users, _someone_ is bound to try that combination :grin: If it means even just one less ticket for our Happiness Engineers, then I thought I might as well cover all the bases :wink:

umurkontaci commented on client/lib/domains/dns/index.js:
```diff
@@ -45,8 +46,8 @@ function isValidDomain( name ) {
-function isValidName( name, type ) {
-	if ( isNameEmptyAndItsAllowed( name, type ) ) {
+function isValidName( name, type, domainName ) {
+	if ( isRootDomainAndItsAllowed( name, type, domainName ) ) {
```
The function name is not totally clear. It's allowed for what?

klimeryk replied:
It's allowed - as in:
![214914](https://cloud.githubusercontent.com/assets/3392497/13216450/1d70254e-d95c-11e5-91dd-d9d68bd527de.gif)
Sorry, couldn't help myself :laughing: 

As in - the opposite of "forbidden". I tried making it read as close to normal language as possible - "if it is a root domain and it's allowed" (aka. "not forbidden"). I honestly can't think of a better word here :/

BandonRandon replied:
What about `isRootDomainAndIsValid`?

klimeryk replied:
I wanted to stress the "allowed" part - the root domain might be in a valid format, but it might not be _allowed_ for the given record type (for example, we allow root domains for TXT but not for CNAME). I wanted the name of the function to hint at that difference: `isRootDomain` means we are dealing with a root domain and `AndItsAllowed` means that we are checking if it's allowed for the given record type.

umurkontaci replied:
`andIsValid` makes more sense, but not entirely. I don't want to stall this PR but it doesn't make sense with "allowed". What does it mean to be allowed? What's the rule behind being allowed? In this case, it seems there are certain types of records that can be root domain. So why not `isRootDomain(name, domainName) && canBeRootDomain(type)`? Or create constant for `ALLOWED_ROOT_DOMAIN_TYPES` and do the check inline?

klimeryk replied:
It's important to check if the input is a root domain _and_ it can be a root domain for the give type. That's why I wanted it to be one function (plus, it's used in two places in that file). But you are right - your version makes it clearer what's what. I'll update the PR :+1:

umurkontaci commented:
Looks good! There's one bit of comment about the function name, but the rest looks good!

klimeryk commented:
Renamed and split `isRootDomainAndItsAllowed` to `isRootDomain` and `canBeRootDomain` :+1:

umurkontaci commented:
LGTM :+1:
